File: Client_C++/prime_client.cpp
#include <iostream>
#include <memory>
#include <string>
#include <grpcpp/grpcpp.h>
#include "prime.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using grpc::ClientReader;
using prime::PrimeService;
using prime::PrimeRequest;
using prime::PrimeResponse;
using prime::FirstPrimeRequest;
using prime::FirstPrimeResponse;
using prime::IntervalRequest;
using prime::PrimeNumber;

class PrimeClient {
private:
    std::unique_ptr<PrimeService::Stub> stub_;

public:
    PrimeClient(std::shared_ptr<Channel> channel)
        : stub_(PrimeService::NewStub(channel)) {}

    // Service 1 : IsPrime
    void IsPrime(long n) {
        PrimeRequest request;
        request.set_number(n);
        
        PrimeResponse response;
        ClientContext context;

        Status status = stub_->IsPrime(&context, request, &response);

        if (status.ok()) {
            if (response.is_prime()) {
                std::cout << n << " est PREMIER." << std::endl;
            } else {
                std::cout << n << " n'est PAS premier." << std::endl;
                std::cout << "Factorisation : ";
                for (int i = 0; i < response.factors_size(); i++) {
                    std::cout << response.factors(i) << " ";
                }
                std::cout << std::endl;
            }
        } else {
            std::cout << "Erreur gRPC: " << status.error_message() << std::endl;
        }
    }

    // Service 2 : FirstPrime
    void FirstPrime(long n) {
        FirstPrimeRequest request;
        request.set_n(n);
        
        FirstPrimeResponse response;
        ClientContext context;

        Status status = stub_->FirstPrime(&context, request, &response);

        if (status.ok()) {
            std::cout << "Les " << n << " premiers nombres premiers :" << std::endl;
            for (int i = 0; i < response.primes_size(); i++) {
                std::cout << response.primes(i) << " ";
            }
            std::cout << std::endl;
        } else {
            std::cout << "Erreur gRPC: " << status.error_message() << std::endl;
        }
    }

    // Service 3 : IntervalPrime (streaming)
    void IntervalPrime(long min, long max) {
        IntervalRequest request;
        request.set_min(min);
        request.set_max(max);
        
        ClientContext context;
        PrimeNumber prime;

        std::unique_ptr<ClientReader<PrimeNumber>> reader(
            stub_->IntervalPrime(&context, request));

        std::cout << "Nombres premiers entre " << min << " et " << max << " :" << std::endl;
        
        int count = 0;
        while (reader->Read(&prime)) {
            std::cout << prime.value() << " ";
            count++;
        }

        Status status = reader->Finish();
        if (status.ok()) {
            std::cout << "\n[Terminé - " << count << " nombres trouvés]" << std::endl;
        } else {
            std::cout << "\nErreur streaming: " << status.error_message() << std::endl;
        }
    }
};

int main() {
    std::string host;
    int port;
    int choix;

    // Saisie de l'adresse du serveur
    std::cout << "Entrez l'IP du serveur : ";
    std::getline(std::cin, host);

    std::cout << "Entrez le port du serveur : ";
    std::cin >> port;
    std::cin.ignore();  // Vider le buffer

    std::string server_address = host + ":" + std::to_string(port);
    std::cout << "Connexion au serveur " << server_address << " ..." << std::endl;

    // Création du canal gRPC
    auto channel = grpc::CreateChannel(server_address, grpc::InsecureChannelCredentials());
    PrimeClient client(channel);

    std::cout << "Connecté !" << std::endl;

    // Menu interactif
    do {
        std::cout << "\n===== MENU =====" << std::endl;
        std::cout << "1. IsPrime       (est-ce que n est premier ?)" << std::endl;
        std::cout << "2. FirstPrime    (les N premiers nombres premiers)" << std::endl;
        std::cout << "3. IntervalPrime (nombres premiers entre Min et Max)" << std::endl;
        std::cout << "0. Quitter" << std::endl;
        std::cout << "Choix : ";
        std::cin >> choix;
        std::cin.ignore();

        long n, min, max;
        switch (choix) {
            case 1:
                std::cout << "Entrez n : ";
                std::cin >> n;
                std::cin.ignore();
                client.IsPrime(n);
                break;
                
            case 2:
                std::cout << "Combien ? ";
                std::cin >> n;
                std::cin.ignore();
                client.FirstPrime(n);
                break;
                
            case 3:
                std::cout << "Min : ";
                std::cin >> min;
                std::cout << "Max : ";
                std::cin >> max;
                std::cin.ignore();
                client.IntervalPrime(min, max);
                break;
                
            case 0:
                std::cout << "Au revoir !" << std::endl;
                break;
                
            default:
                std::cout << "Choix invalide." << std::endl;
        }
    } while (choix != 0);

    return 0;
}

File: ClientJava/src/main/java/com/example/grpcclient/PrimeClient.java
package com.example.grpcclient;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;
import prime.PrimeServiceGrpc;
import prime.Prime.*;

import java.util.Scanner;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class PrimeClient {

    private final PrimeServiceGrpc.PrimeServiceBlockingStub blockingStub;
    private final PrimeServiceGrpc.PrimeServiceStub asyncStub;

    public PrimeClient(String host, int port) {
        ManagedChannel channel = ManagedChannelBuilder
                .forAddress(host, port)
                .usePlaintext()
                .build();
        blockingStub = PrimeServiceGrpc.newBlockingStub(channel);
        asyncStub = PrimeServiceGrpc.newStub(channel);
    }

    // Service 1 : IsPrime - requête/réponse normale
    public void isPrime(long n) {
        PrimeResponse res = blockingStub.isPrime(
                PrimeRequest.newBuilder().setNumber(n).build()
        );
        if (res.getIsPrime())
            System.out.println(n + " est PREMIER.");
        else {
            System.out.println(n + " n'est PAS premier.");
            System.out.println("Factorisation : " + res.getFactorsList());
        }
    }

    // Service 2 : FirstPrime - requête/réponse normale
    public void firstPrime(long n) {
        FirstPrimeResponse res = blockingStub.firstPrime(
                FirstPrimeRequest.newBuilder().setN(n).build()
        );
        System.out.println("Les " + n + " premiers nombres premiers :");
        System.out.println(res.getPrimesList());
    }

    // Service 3 : IntervalPrime - streaming serveur → client
    public void intervalPrime(long min, long max) throws InterruptedException {
        CountDownLatch latch = new CountDownLatch(1);
        asyncStub.intervalPrime(
                IntervalRequest.newBuilder().setMin(min).setMax(max).build(),
                new StreamObserver<PrimeNumber>() {
                    public void onNext(PrimeNumber v)  { System.out.print(v.getValue() + " "); }
                    public void onError(Throwable t)   { System.err.println(t.getMessage()); latch.countDown(); }
                    public void onCompleted()          { System.out.println("\n[Terminé]"); latch.countDown(); }
                }
        );
        latch.await(30, TimeUnit.SECONDS);
    }

    public static void main(String[] args) throws InterruptedException {
        Scanner sc = new Scanner(System.in);

        // Demander l'IP et le PORT du serveur
        System.out.print("Entrez l'IP du serveur : ");
        String host = sc.nextLine().trim();

        System.out.print("Entrez le port du serveur  : ");
        int port = sc.nextInt();
        sc.nextLine(); // vider le buffer

        System.out.println("Connexion au serveur " + host + ":" + port + " ...");
        PrimeClient client = new PrimeClient(host, port);
        System.out.println("Connecté !\n");

        int choix;
        do {
            System.out.println("\n===== MENU =====");
            System.out.println("1. IsPrime       (est-ce que n est premier ?)");
            System.out.println("2. FirstPrime    (les N premiers nombres premiers)");
            System.out.println("3. IntervalPrime (nombres premiers entre Min et Max)");
            System.out.println("0. Quitter");
            System.out.print("Choix : ");
            choix = sc.nextInt();

            switch (choix) {
                case 1 -> { System.out.print("Entrez n : "); client.isPrime(sc.nextLong()); }
                case 2 -> { System.out.print("Combien ? "); client.firstPrime(sc.nextLong()); }
                case 3 -> {
                    System.out.print("Min : "); long min = sc.nextLong();
                    System.out.print("Max : "); client.intervalPrime(min, sc.nextLong());
                }
                case 0 -> System.out.println("Au revoir !");
                default -> System.out.println("Choix invalide.");
            }
        } while (choix != 0);
    }
}
